zilliqa/src/consensus.rs

```diff
@@ -1903,37 +1953,80 @@ impl Consensus {
     }
 
     pub fn leader_at_block(&self, block: &Block, view: u64) -> Option<Validator> {
+        if let Some((cached_block_num, cached_view, leader)) = *self.leader_cache.borrow() {
+            if cached_block_num == block.number() && cached_view == view {
+                return Some(leader);
+            }
+        }
+
+        let Ok(state_at) = self.try_get_state_at(block.number()) else {
+            return None;
+        };
+        let leader = self.leader(&state_at, view);
+
+        Some(leader)
+    }
+
+    pub fn leader_at_block_bytes(&self, block: &Block, view: u64) -> Option<(Vec<u8>, PeerId)> {
         let Ok(state_at) = self.try_get_state_at(block.number()) else {
             return None;
         };
-        Some(self.leader(&state_at, view))
+        let leader = self.leader_bytes(&state_at, view);
+
+        Some(leader)
+    }
+
+    pub fn leader_bytes(&self, state: &State, view: u64) -> (Vec<u8>, PeerId) {
+        let committee = state.get_stakers_bytes().unwrap();
+
+        let mut rng = ChaCha8Rng::seed_from_u64(view);
+        let dist = WeightedIndex::new(committee.iter().map(|pub_key| {
+            let stake = state
+                .get_stake_from_public_key_bytes(pub_key.clone())
+                .unwrap()
+                .context("Committee member has no stake")
+                .unwrap();
+            stake.get()
+        }))
+        .unwrap();
+        let index = dist.sample(&mut rng);
+        let public_key = committee.get(index).unwrap();
+
+        let peer_id = self
+            .state
+            .get_peer_id_from_public_key_bytes(public_key.clone())
+            .unwrap()
+            .context("Unable to get peer_id from staking contract!")
+            .unwrap();
+
+        (public_key.clone(), peer_id)
     }
 
     pub fn leader(&self, state: &State, view: u64) -> Validator {
-        let committee = state.get_stakers().unwrap();
+        let committee = state.get_stakers_bytes().unwrap();
 
         let mut rng = ChaCha8Rng::seed_from_u64(view);
         let dist = WeightedIndex::new(committee.iter().map(|pub_key| {
             let stake = state
-                .get_stake(*pub_key)
+                .get_stake_from_public_key_bytes(pub_key.clone())
                 .unwrap()
                 .context("Committee member has no stake")
                 .unwrap();
             stake.get()
         }))
         .unwrap();
         let index = dist.sample(&mut rng);
-        let public_key = *committee.get(index).unwrap();
+        let public_key = committee.get(index).unwrap();
 
         let peer_id = self
             .state
-            .get_peer_id(public_key)
+            .get_peer_id_from_public_key_bytes(public_key.clone())
             .unwrap()
             .context("Unable to get peer_id from staking contract!")
             .unwrap();
 
```

Comment:
This seems shared with `leader_bytes()` above. If I'm not missing any details, it should be possible to reduce `leader()` to something along the lines of
```rust
pub fn leader(&self, state: &State, view: u64) -> Validator {
    let (pub_key_bytes, peer_id) = self.leader_bytes(state, view);
    Validator {
        public_key: NodePublickey::from_bytes(pub_key).unwrap(),
        peer_id,
    }
}
```